### DISCUSSION TITLE
Everything becomes slow in VS Code when Copilot is enabled

### DISCUSSION THREAD

**Discussion by from-nibly** [community/community#10053] [Copilot Conversations]
So recently whenever I turn on Copilot, everything in my vscode starts slowing down. Intelligent never resolves, linting errors take forever, and basically all features outside of opening files and typing in them directly take a really long time to resolve.

Interestingly enough it doesn't even seem to help if I disable it for just typescript. It only get's fixed if I disable the extension itself.

Is anyone else having this issue? 

is there a good way to peek into what's going on?

**Comment by ultmaster** [comment]
Same issue here starting around days ago. My suspicion is that some recent updates have broken things. Haven't figured it out really. I had to disable the extension temporarily.

**Reply by Dev-JosephSuvak** [reply]
I think they have even broken the token counter because it'll literally freeze on a simple task, even though my tokens keep increasing exponentially. Not sure how that works.

**Comment by alexvoedi** [comment]
Same problem in small python project.

**Comment by vdods** [comment]
Same problem in Rust and Python projects.  Even saving a file can take up to 7 seconds to complete.  It makes the CPU chug very hard for a surprisingly long time even for a couple edits here and there.  I can reliably fix this by disabling the copilot extension, and it reliably reoccurs when re-enabling the copilot extension.

**Comment by Oba-One** [comment]
Same issue and sketches me out a little ngl

**Reply by gitHubhuanhuan** [reply]
hi ，Can I borrow your copilot? I can't use my student bag, thank you，Maybe you can activate it for me

**Comment by dan-cooke** [comment]
Was working just fine for weeks, and now all of a sudden on a new Remix project it is slowing to a halt. I'm wondering if its parsing node_modules or something by mistake.

**Reply by gitHubhuanhuan** [reply]
hi ，Can I borrow your copilot? I can't use my student bag, thank you，Maybe you can activate it for me

**Comment by asif17r** [comment]
I have the same issue here with C++

**Reply by gitHubhuanhuan** [reply]
hi ，Can I borrow your copilot? I can't use my student bag, thank you，Maybe you can activate it for me

**Comment by Dredok** [comment]
same here, also with C++ project (on ubuntu and using latest vscode, if that matters). 
disabling the extension everything works as smoothly as before ...

**Comment by dougdyson** [comment]
Copilot has been really frustrating the developer experience the past couple of days - going to explore alternatives, as there are other options now.

**Reply by gitHubhuanhuan** [reply]
hi ，Can I borrow your copilot? I can't use my student bag, thank you，Maybe you can activate it for me

**Comment by ZupoLlask** [comment]
I'm noticing the following pattern while working with Jupyter notebooks: when things get slow, Copilot status icon is spinning.

Maybe some API is taking too much to return some answer or some API background service is crashing...

**Reply by codinginflow** [reply]
> I'm noticing the following pattern while working with Jupyter notebooks: when things get slow, Copilot status icon is spinning.

Same problem here and I noticed the same behavior.

**Comment by Recodify** [comment]
Same here. It's so slow that I've had to disable it. Come guys, this is a paid service and it's currently, for all practical purposes, broken

**Reply by gitHubhuanhuan** [reply]
hi ，Can I borrow your copilot? I can't use my student bag, thank you，Maybe you can activate it for me

**Comment by lebifeng** [comment]
Check your activity monitor then maybe you will find out that some process is running with full cpu usage.

**Comment by vdods** [comment]
FYI, I switched to VSCode Insiders edition and this is no longer a problem.  Copilot works and doesn't slow anything down.

**Reply by ZupoLlask** [reply]
What is the oldest Insiders version you think fixed your issue, or when/how long did you switch to Insiders?

**Reply by vdods** [reply]
I don't know.  I installed it initially about 2 weeks ago.

**Reply by ZupoLlask** [reply]
Hi @vdods,

So how's your feedback 3 weeks switching to Insider's? Do you still consider the issue as fixed from your experience?

**Reply by vdods** [reply]
Yeah, it still works well.

**Reply by ZupoLlask** [reply]
Version 1.80.*, right?

**Reply by codinginflow** [reply]
I use Insiders and I still get the hang regularly.

**Reply by vdods** [reply]
Yes, 1.80.* works for me.

**Comment by AbdullahWins** [comment]
As copilot constantly goes through your code to make sense out of it, it is expected that it'll slow your computer down a bit as it's constantly running on background while you're using vscode.

You can minimize the workload by removing unused background tasks or maybe upgrade the computer if needed.

**Reply by codinginflow** [reply]
> As copilot constantly goes through your code to make sense out of it, it is expected that it'll slow your computer down a bit as it's constantly running on background while you're using vscode.
> 
> You can minimize the workload by removing unused background tasks or maybe upgrade the computer if needed.

I have a high-end PC and it still becomes crippling slow. It can be mitigated by reloading VS Code so there must be some underlying bug.

**Reply by AbdullahWins** [reply]
Yeah that's another possibility, maybe it's triggering multiple instances which is slowing the vscode down.
It's safe to say maybe they broke something in this update

**Reply by gitHubhuanhuan** [reply]
hi ，Can I borrow your copilot? I can't use my student bag, thank you，Maybe you can activate it for me

**Comment by codinginflow** [comment]
I have the same problem. I found that I can improve the speed again by executing "Developer: Reload Window" in VS Code. Then once it becomes slow again, I repeat it. 

Obviously not a great longterm solution but I'm sure this will be fixed eventually.

**Comment by biggrizzleyguy** [comment]
Any fix for this yet?  Autopilot causing my Visual Studio 2023 run slow on my high end machine.

**Comment by asif17r** [comment]
For me, the problem fixed up to 95% when I disabled "C/C++" extension. Now I only have "code runner" and "github copilot" extension. You guys can try this fix if it works.

**Comment by Hupdy** [comment]
Nice

**Comment by abdul-alhasany** [comment]
I have the same issue. While working on a PHP linting extension I noticed that formatting the document is taking a very long time.

When debugging the issue I noticed that `workspace.onDidChangeTextDocument` is being triggered constantly without stop. This occurs even when no text is being changed.

You can reproduce the issue by creating an empty extension and add `console.log` or use output channel log inside the event.

**Reply by adhal007** [reply]
I had this one extension called linter which was causing this issue - was able to narrow down to this and disable it for better performance.

**Comment by nurgazik** [comment]
I get that copilot might be going through the codebase and analyzing the code, thus it's responses might be slow, but my entire VS Code is slow. When I am typing something into the chat window the rendering of characters lag for an entire second or so, making it almost unusable. I don't have the most powerful PC in the world, but it's still a macbook pro core i7 16GB ram. Does anyone experience the same?

**Reply by jaakko-paavola** [reply]
My VSCode was close to unusable with Copilot still a couple weeks ago. I did some searching and tweaking and now it is significantly better. Two things that I believe made all the difference:
1) I stopped working with a folder open, i.e., with a workspace, only opening individual files (and keeping as few of them open as possible).
2) I keep disabling Copilot with a key shortcut when I don't need it, because the worst lags coincide with the little Copilot icon in the bottom right corner spinning as Copilot is fetching an inline suggestion. The key shortcut required a little hack using the Toggle extension, because the shortcut "enable/disable Copilot completions" does not seem to work. If people could go upvote my [bug report about this](https://github.com/orgs/community/discussions/79048) would be good to get it some visibility.

**Reply by JeBehen** [reply]
My IntelliJ has become very slow since Copilot was introduced. All of my apps are slow now.

**Reply by packley1** [reply]
My IntelliJ has become very slow too (since Copilot).  Noticed too that my (Windows) laptop is much slower as well.  Have there been any solutions for this?

**Comment by grindstm** [comment]
Working in Jupyter notebooks is becoming impossible. Most of the time, there are full seconds of delay between typing and characters appearing on the screen, particularly in markdown cells. The keyboard shortcut for disabling copilot completions isn't working as expected. I have tried the insiders release of vs code and the pre-release of copilot to no avail. This problem has only recently gotten insanely bad. Pylance and copilot also seem to be fighting each other.

**Reply by samraul** [reply]
Agreed. Sometimes starting to run a notebook cell takes an entire minute, even though the cell itself takes less than a second to run.

**Reply by chrisoffner3d** [reply]
I have the same problem. I'm running VS Code 1.87.0 on an M3 Max MacBook Pro, and using Jupyter Notebooks with Python kernels while having the GitHub Copilot extension on makes everything **very** slow and laggy.

**Reply by Jeffresh** [reply]
Same problem here, and it's strange because it doesn't happen until you run a bunch of cells and stay in them for a while. If you close the project and reopen it, vscode runs fine until you spend some time working.

It is like a memory bug or something. It doesn't happen when working with flat Python code.

**Reply by JasonGross** [reply]
It happens when working with Jupyer view on flat Python code, see https://github.com/orgs/community/discussions/121204

**Comment by JeBehen** [comment]
Same problem in a small JavaFX project. I thought my IntelliJ was acting up. I opened jGrasp and a dialog box opened and said jGrasp opened  to slowly and suggested I up grade my video driver.
![image](https://github.com/community/community/assets/129677019/6e844114-3502-4ea7-a84c-aef652fe50fe)

**Reply by gitHubhuanhuan** [reply]
hi ，Can I borrow your copilot? I can't use my student bag, thank you，Maybe you can activate it for me

**Comment by JasonGross** [comment]
I've reported a more specific issue at https://github.com/orgs/community/discussions/121204, where "Run cell" in Jupyter view of .py files in VSCode takes many seconds before actually running the code.

**Comment by bnrosa** [comment]
Same issue here sometimes it take seconds to save a file, and I'm using a really potent PC (12th gen Intel Core i7, 64GB of RAM, etc). It would be great if we could at least disable the copilot actions that run on file save.

**Comment by RexBarker** [comment]
I was having the same issue.  For large projects, it seems copilot is too ambitious with the analysis.  You can try reducing the scope of the search by changing the file settings (`settings.json` (USER)):
![image](https://github.com/community/community/assets/22831637/cf9e8ffa-180c-45b6-a281-f7b460e8ee18)

**Reply by lucasscurtoo** [reply]
Hey, i did that and doesn't worked, the thing with the issue is i can't use ctrl + space ( don't know the name sorry) i have to reload vs code and if i open the chat it starts the problem

**Reply by gitHubhuanhuan** [reply]
hi ，Can I borrow your copilot? I can't use my student bag, thank you，Maybe you can activate it for me

**Reply by Sandbye** [reply]
> Hey, i did that and doesn't worked, the thing with the issue is i can't use ctrl + space ( don't know the name sorry) i have to reload vs code and if i open the chat it starts the problem

This is the same problem i have, it starts slowing vscode down, when opening the chat (loading it)

**Reply by HadenHiles** [reply]
This made intellisense way faster for me again. Thank you!!

Definitely an issue that needs to be addressed but this is a welcomed workaround for now.

**Reply by RustyLesh** [reply]
1 year later, and this still helped a lot with my lag.

**Comment by emok032** [comment]
Same issue. Seems like Intellisense is not as responsive anymore. Can't get any Intellisense feedback or suggestions or it's extremely slow like it's waiting for Copilot to finish something (even though it's not doing anything at the moment).

**Reply by JeBehen** [reply]
Go to your settings and disable Copilot.

**Reply by gitHubhuanhuan** [reply]
hi ，Can I borrow your copilot? I can't use my student bag, thank you，Maybe you can activate it for me

**Comment by gitHubhuanhuan** [comment]
hi ，Can I borrow your copilot? I can't use my student bag, thank you，Maybe you can activate it for me

**Comment by Sandbye** [comment]
Have same issue on Linux Ubuntu 22.04 ... having copilot activated slows my  whole vscode

**Comment by Sandbye** [comment]
For people that have a large chat, try and create a new one by pressing the + button in vscode:

![image](https://github.com/user-attachments/assets/c4c3e4b1-b60d-47ad-a7c7-829341ff9550)

I have a suspicion, that maybe copilot can't handle a chat that's too big, or that it generates something that takes a long time to analyze.

This "fix", worked for me

**Reply by isaac-svg** [reply]
I think it indexes the files so the more chat and files you have more the index gets bigger.

**Reply by isaac-svg** [reply]
Your solution works for thanks.

**Reply by YiLiangBJ** [reply]
This works for me. Thank you very much.

**Comment by evgeniiarts** [comment]
My VS Code is unusable for typescript when enabling Copilot. 
I have a typescript project and type checks do not work properly with enabled Copilot. A type check can take several minutes to show an error, if at all.
While the same type checks work flawlessly and instantly when Copilot is either disabled.

I have **disabled all extensions** for testing this, and can confirm that Copilot alone can mess up your workflow badly.

**Reply by san4d** [reply]
Do you find a work-around for this? I'm running into this with a medium size app.

**Reply by evgeniiarts** [reply]
Sadly no. 

I have tried increasing memory allocation within my win11 system and additionally within VSCode without success.
My IDE keeps getting stuck on "Loading IntelliSense status...".
After spending hours reinstalling VS Code, cleaning caches etc. I decided that Copilot isn't worth the time investment anymore. 

I opted for laying off co-pilot in the end.

**Reply by aslaker** [reply]
I'm running into this same issue.  The intellisense status for my Typescript never resolves, and the only way to fix it is to disable co-pilot.  I have tried removing all other plugins to see if there is another conflict, and it is only co-pilot that causes this for me.  I'm running MacOS Sonoma, so it appears based on your comment that it is not system specific.

For the time being, I've opted to not use co-pilot even after being a heavy user for a while.

**Reply by fahimShahriaar** [reply]
I'm also facing the same issue. The file saving process is also very very slow.

**Reply by evgeniiarts** [reply]
i have recently purchased a completely new pc and freshly installed win11, did not keep anything from my old system. 96gb RAM and a new AMD processor, so the hardware is definitely not the issue.

freshly installed VSCode and Copilot as the only extension. opened a small typescript project - and the same issue persists.
IntelliSense loading forever.. incredible.

**Comment by JeBehen** [comment]
Uninstall CoPilot, allocate more virtual memory to your system.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by heytulsiprasad** [comment]
Damn 2025 already, we're still here. 🥲

**Reply by kulcsarrudolf** [reply]
Is there any solution? Worked fine for me until last week.

**Reply by kulcsarrudolf** [reply]
Also works very smoothly when I am using it on my Mac.

**Comment by Shivansh-Verma7719** [comment]
Any solution for this yet, this really drains my battery as well. Even Prettier slows down.

**Comment by zengjunlin2022** [comment]
Yesterday, when I saved my changes, it prompted that the formatting process was running and took over a minute to save a single file, which is unusual. However, once I disabled Copilot Chat, everything returned to normal. The problem reappeared as soon as I re-enabled the Copilot Chat extension, even if I only changed a single character. But today, when I reactivated the Copilot Chat extension again, it worked normally, which left me puzzled. What could be causing this?

**Reply by JPHutchins** [reply]
This is what I am observing as well. I've lost time to compiling, only to realize that the save NEVER completed, so I'll need to exit and recompile after letting the save finish. Which can take seconds, where I expect milliseconds.

I'll add that cutting and pasting code occasionally results in a 1-2 second spinner that I'd never seen before.

It's like the editor is blocking on Copilot I/O, or Copilot is fighting the other on-<action> formatters/linters!

The problems do go away by disabling copilot.

**Comment by jasonwheritage** [comment]
I have the same issue. Initially performance is ok, but 5 minutes into my session and the following things take forever: ctrl-click navigation, copilot suggestions/autocomplete. Chat responses sometimes degrade to the point they are no longer useable. Starting a new chat or restarting vscode helps with this last issue.

**Comment by isaac-svg** [comment]
Same issue here VS Code gets too slow when I switch to agentic mode but works fine when I am in ask mode.
Ubuntu 25

**Comment by HadenHiles** [comment]
Same issue here in a Flutter project. Can't get any of the "quick fixes" to load while copilot extension is active. I thought that the copilot extension was supposed to handle this by only trying to understand the file that is in context...

There's got to be a way to only activate copilot's background processes when it's actually asked to perform a task - at that point I don't mind if it's a little bit slower.

Disappointing this is still an issue!

**Comment by UmiKami** [comment]
Same issue with a large flask python project.

**Comment by sskserk** [comment]
Same issue

**Comment by jacastillo302** [comment]
Same issue with  Go proyect

**Comment by holliape** [comment]
closing some of the many files I had open helped copilot go waaaay faster.

**Comment by yaser-alazm** [comment]
Same issue, strange thing that it still not fixed!

**Comment by xNepenthe** [comment]
2025 and this is still a thing, even pasting something (ctrl + v) lags.

**Reply by jesse-beckton-alight** [reply]
And I thought it was just me, I've been fighting with this for many weeks now trying to figure it out. I just disable Copilot all together.

**Comment by garyjkuehn** [comment]
With the latest drop of VSCode Version: 1.104.2 performance fell off the cliff thanks to everything AI. Since the trend seems to be permeate AI thru everything, users be damned, I finally disabled and hid all AI features provided by GitHub CoPilot via settings.  Now if I want AI assistance I use alternate means.  Now I have back a responsive IDE with lightening fast linting (ruff) and itellisense that is enjoyable and certainly responsive.  Granted it would be nice to use AI on demand in the IDE but its current implementation and I suspect follow-on will be more feature bloat to the detriment of performance.  No, thank you.

**Comment by jesse-beckton-alight** [comment]
I've had to resort to disabling Copilot as it destroys my vs code experience. Hoping to switch to Claude Code soon.

**Comment by daviddanielng** [comment]
How can the vscode devs prioritise copilot over every other things, no doubt AI is helpful, now it is just impending . 
I want copilot, but i want my autocomplete prioritised over copilot.

**Comment by JE-lee** [comment]
I'm using Windows 10 with WSL. When I'm using GitHub Copilot Chat on a large project (perhaps 2 million lines of code), after one or two conversations, the entire VSCode system almost inevitably becomes very slow, including switching tabs, typing, and so on. I've also observed that during this period of lag, all my Chromium-based programs will even flash black screens. Gemini tells me that a problem with my graphics card driver or insufficient memory is likely the cause. I tried closing many other software programs and ensuring my computer's memory load is relatively low (<70%), and everything has improved. Copilot works quickly, and VSCode runs smoothly. The above is for reference only.

**Comment by MathanGeurtsen** [comment]
I'd just like to add a voice to the choir: I've just disabled copilot, and the dreadful performance improved somewhat. Like everyone else here, I'm not going to use copilot if it means hamstringing the rest of vscode's features. 

(Interesting to note: I'm seeing no such problems in the excellent `github/copilot.vim` plugin I'm using in neovim, it's a joy to use there!)

**Comment by kristiangasic** [comment]
Quick fix for GitHub Copilot Chat in vscode:

Open this path in your Explorer:
C:\Users\YourUSER\AppData\Roaming\Code\User\globalStorage\github.copilot-chat

Delete the file: **commandEmbeddings.json**

Restart VSCode or your PC – done!

**Reply by sandeepbanala** [reply]
Hey there! 

Doing that wouldn't delete the chat context?

**Reply by kristiangasic** [reply]
No, deleting the commandEmbeddings.json file does not delete your chat context or chat history :)
file stores embeddings for VSCode commands that GitHub Copilot Chat uses to better understand VSCode-specific commands. It's automatically created during initial indexing and contains no personal chat data

**Reply by Mathews-axisTech** [reply]
This seemed to do the trick!
I also didn't go to open any previous chat contexts after deleting the file. And that fixed the issue for me.

**Comment by 00xDisaster** [comment]
## Everything becomes slow in VS Code when Copilot is enabled (#10053)

### Problem
When Copilot is enabled, VS Code becomes very slow:
- IntelliSense never resolves
- Linting and saving takes many seconds
- Typing and navigation lag
- Performance returns to normal only when Copilot is disabled

This happens across different languages (TypeScript, Python, Rust, C++, etc.) and persists for many users.

---

## Why this happens (Root Cause)
Copilot runs background analysis and indexing for your project.  
When the project is large or has many open files, Copilot can:
- consume high CPU/RAM
- run continuous background processes
- interfere with IntelliSense and other extensions
- cause VS Code to freeze or lag

Many users reported that the slowdown **starts after some time**, and reloading VS Code fixes it temporarily.

---

## Common Observations from Users
- Slowdowns occur especially in:
  - Jupyter notebooks
  - large TypeScript projects
  - large Python projects
  - projects with many files
- Some users fix it by:
  - disabling Copilot
  - reloading VS Code window
  - using VS Code Insiders (some versions fixed it)
  - closing many open files
  - reducing Copilot file scope

---

## Workarounds (Most Effective)
### ✅ 1. Disable Copilot when not needed
This is the most reliable workaround.

### ✅ 2. Reload Window
Developer: Reload Window
This temporarily restores performance.

### ✅ 3. Reduce Copilot Scope
Add this in `settings.json`:

```json
"copilot.fileExtensions": [
  "js",
  "ts",
  "py"
],
"copilot.enable": true

```
Or reduce workspace indexing using:

```json
"copilot.exclude": [
  "**/node_modules/**",
  "**/.git/**",
  "**/dist/**"
]
```

### ✅ 4. Close many open files / don’t open workspace

Users reported better performance by:

Opening single files instead of a folder

Keeping few tabs open

### ✅ 5. Delete Copilot Chat embeddings (if using Copilot Chat)

Path:

    %APPDATA%\Code\User\globalStorage\github.copilot-chat


Delete:

    commandEmbeddings.json


Then restart VS Code.

---

### Why this is still not fixed

Many users have reported this issue since 2022 and it continues to appear in 2025.
It may be due to:

 - Copilot’s background indexing

 - Conflicts with other extensions

 - VS Code performance changes

 - Large project size

---

### Summary

If Copilot slows down VS Code:

 - It’s not a single bug — it’s related to background indexing and project size

 - The most reliable fix is to disable Copilot

 - Temporary fixes: reload window, close files, reduce scope

 - If using Copilot Chat, deleting commandEmbeddings.json often helps

---

### Sources

 - GitHub Community discussion (#10053)

 - User reports across 2022–2025 (multiple comments)

 - GitHub Copilot documentation

**Reply by towfiqi** [reply]
A `"copilot.exclude"` does not exist. Please stop copy-pasting hallucinated answers from AI.

**Comment by Sandbye** [comment]
I had this issue a long time ago, and it hasn't come up again.
Make sure VS Code is up to date, and delete old big chat history.
If VS Code is slowing down, most likely the issue stems from having a large chat.

**Comment by MrBns** [comment]
This issue is here for a long time. but still marked as Unanswered. 
I am still having similer problem.

**Comment by AdjWang** [comment]
Same issue with C++, it even slows down typing to an untoleratable level. Restart vscode or reboot does not help, the only way is to disable Copilot.

**Comment by luho** [comment]
Deleting my copilot-instructions.md file fixed the issue for me.

**Comment by Parveen012** [comment]
I’ve faced something similar with Copilot slowing down VS Code, especially when working with TypeScript. In my case, the issue wasn’t just Copilot itself but how it interacted with the TypeScript server and other extensions.

A few things that helped me debug it:

Open Command Palette → “Developer: Show Running Extensions”
→ This shows which extension is consuming the most time/CPU.
Check “Developer: Open Process Explorer”
→ You can see if the TypeScript server or extension host is overloaded.
Try running VS Code with extensions disabled (--disable-extensions) and then enable only Copilot
→ Helps confirm if it’s a conflict with another extension (ESLint, Prettier, etc.)
Also check the Output panel → GitHub Copilot logs
→ Sometimes it shows delays or request issues.

In my case, the slowdown improved after:

Disabling unnecessary extensions
Restarting the TypeScript server
Updating VS Code + Copilot to the latest version

If disabling Copilot completely fixes it for you, it’s likely either:

Extension conflict
High resource usage from Copilot + TS server together

Would be interesting to know your project size and what other extensions you’re running — that can make a big difference.

**Reply by MrBns** [reply]
>Disabling unnecessary extensions
Restarting the TypeScript server
Updating VS Code + Copilot to the latest version

bro, there is a reason people use extension.

**Comment by stan-at-work** [comment]
Same

**Comment by Arteiimis** [comment]
Will this issue ever be resolved?

**Comment by gamlinskijoleg** [comment]
Still getting this issue.

**Comment by yousefbinali** [comment]
This is still happening. Specially with C++ IntelliSense. Whenever copilot is enabled it takes ages (sometimes never) to finish scanning.

**Comment by n-ii-ma** [comment]
Same issue.
